feat: switch to external Rust-based language server

lemonadern opened the pull request:
## Summary

Rust 製言語サーバに対応する差分です
挙動担保のため E2E テストを追加しています

## コマンドについて

### 変更前：
全てを`Format SQL`（`uroborosql-fmt.uroborosql-format`）でフォーマットする
- 選択箇所があれば選択箇所を SQL としてフォーマットする
- 選択箇所がなければドキュメント全体を SQL としてフォーマットする

### 変更後：
`Format SQL` コマンドの挙動に変更無し（どんなドキュメントでも、選択箇所あるいはドキュメント全体を SQL としてフォーマットを試みる）

そのうえで次が可能になります：
1. SQL ファイルについて
    - VS Code 標準コマンドである `Format Document`（ドキュメント全体のフォーマット）, `Format Selection`（選択箇所のフォーマット）を利用
2. SQL 以外のファイルについて
    - `Formt Selection as SQL` コマンドを新しく提供、選択箇所が SQL としてフォーマットされる

## 実装について
- フォーマット処理は完全に言語サーバを利用した方法に移行しており、サーバ起動以外の napi-rs エンドポイントを利用していません。
- LSP 標準の機能で賄えない機能についてはカスタムリクエストを利用して実現しています
    - 仕様など： https://github.com/future-architect/uroborosql-fmt/blob/main/crates/uroborosql-language-server/docs/protocol.md

## E2E テスト or 動作確認の方法

以下を実施するとローカルで動作確認ができます（煩雑ですが）：
  1. vscode-uroborosql-fmt の依存導入
  2. postinstall script の回避（配布版を使わず、ローカルビルドした napi crate を利用するため）
  3. uroborosql-fmt-napi のローカル build
  4. ビルド成果物を使った tarball の作成
  5. vscode-uroborosql-fmt/server への tarball 差し替え
  6. vscode-uroborosql-fmt の npm run compile

<details>
<summary>（参考）以上の内容を実施するスクリプト</summary>

前提：
```
- （スクリプトを実行するディレクトリ）
    - prepare_embedded_sql_review_env.sh
    - uroborosql-fmt/ (branch: feat/embedded-sql-format-request)
    - vscode-uroborosql-fmt/ (branch: feat/linter-included-language-server)
```

`prepare_embedded_sql_review_env.sh`：
```prepare_embedded_sql_review_env.sh 
#!/usr/bin/env bash

set -euo pipefail

SCRIPT_DIR="$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)"

UROBORO_FMT_REPO="${UROBORO_FMT_REPO:-$SCRIPT_DIR/uroborosql-fmt}"
VSCODE_UROBORO_REPO="${VSCODE_UROBORO_REPO:-$SCRIPT_DIR/vscode-uroborosql-fmt}"

die() {
  echo "error: $*" >&2
  exit 1
}

require_dir() {
  local dir="$1"
  [[ -d "$dir" ]] || die "directory not found: $dir"
}

require_cmd() {
  local cmd="$1"
  command -v "$cmd" >/dev/null 2>&1 || die "command not found: $cmd"
}

require_cmd npm
require_cmd ls
require_cmd cp
require_cmd mkdir
require_cmd rm
require_cmd git

require_dir "$UROBORO_FMT_REPO"
require_dir "$VSCODE_UROBORO_REPO"
require_dir "$UROBORO_FMT_REPO/crates/uroborosql-fmt-napi"
require_dir "$VSCODE_UROBORO_REPO/client"
require_dir "$VSCODE_UROBORO_REPO/server"

FMT_REPO="$UROBORO_FMT_REPO"
VSCODE_REPO="$VSCODE_UROBORO_REPO"
NAPI_DIR="$FMT_REPO/crates/uroborosql-fmt-napi"
PKG_DIR="$NAPI_DIR/repo-review"
SERVER_DIR="$VSCODE_REPO/server"

echo "Using repositories:"
echo "  uroborosql-fmt:          $FMT_REPO"
echo "  vscode-uroborosql-fmt:   $VSCODE_REPO"
echo
echo "Current branches:"
echo "  uroborosql-fmt:          $(git -C "$FMT_REPO" branch --show-current || true)"
echo "  vscode-uroborosql-fmt:   $(git -C "$VSCODE_REPO" branch --show-current || true)"
echo

echo "[1/4] Installing vscode-uroborosql-fmt root/client dependencies without root postinstall"
(
  cd "$VSCODE_REPO"
  npm install --ignore-scripts
  npm install --prefix client --ignore-scripts
)

echo "[2/4] Building uroborosql-fmt-napi and packing a local tarball"
(
  cd "$NAPI_DIR"
  npm run build
  rm -rf "$PKG_DIR"
  mkdir -p "$PKG_DIR"
  cp index.js index.d.ts package.json ./*.node "$PKG_DIR"/
  cd "$PKG_DIR"
  npm pack >/dev/null
)

TARBALL="$(ls -t "$PKG_DIR"/uroborosql-fmt-napi-*.tgz | head -n 1)"
[[ -f "$TARBALL" ]] || die "packed tarball not found under $PKG_DIR"
SERVER_TARBALL_NAME="$(basename "$TARBALL")"

echo "[3/4] Installing local tarball into vscode server"
(
  cd "$SERVER_DIR"
  cp "$TARBALL" "$SERVER_TARBALL_NAME"
  npm install --ignore-scripts --package-lock=false
)

echo "[4/4] Compiling vscode-uroborosql-fmt"
(
  cd "$VSCODE_REPO"
  npm run compile
)

cat <<EOF

Preparation completed.

Local tarball:
  $TARBALL

Next step:
  cd "$VSCODE_REPO"
  npm test

Important:
  Do not run plain 'npm install' at the vscode-uroborosql-fmt root after this.
  Its postinstall downloads the default napi package and can overwrite this local setup.
EOF
```

</details>


ここまでが終われば以下が可能になります：
  1. vscode-uroborosql-fmt へ移動
  2. 動作確認
      - `npm test` で E2E テストを実行（※WSL で動作を確認しています）
      - VS Code を開き Development host を起動して手動確認（※MinGW で動作を確認しています）

lemonadern committed:
feat: migrate to language server mode

Start Rust language server via runLanguageServer() (Node API) and
connect it with vscode-languageclient over stdio. SQL files now use
LSP for Format Document, Format Selection, and save-triggered lint
diagnostics. Add lintConfigurationPath setting and fix Format SQL
command napi import and config path resolution.

lemonadern committed:
feat: migrate to language server mode

Start Rust language server via runLanguageServer() (Node API) and
connect it with vscode-languageclient over stdio. SQL files now use
LSP for Format Document, Format Selection, and save-triggered lint
diagnostics. Add lintConfigurationPath setting and fix Format SQL
command napi import and config path resolution.

lemonadern committed:
test(client): add integration tests for formatting and diagnostics

Add VS Code integration tests that verify Format Document, Format
Selection, and save-triggered lint diagnostics work correctly through
the language server. Also update @vscode/test-electron to ^2.4.0 and
add test fixture SQL files.

lemonadern committed:
refactor: remove client formatter command path

lemonadern committed:
feat: add embedded SQL format command

Add the embedded SQL format command and fold in the related test, fixture, naming, and cleanup follow-ups.

lemonadern committed:
Feat/embedded sql format command (#105)

* feat(client): add format selection as sql command

* style(client): run prettier

* chore(server): drop obsolete napi formatter typings

* test(command): clarify embedded sql test naming

* feat(client): restore format sql command

---------

Co-authored-by: Taishi Naka <lemonadern@users.noreply.github.com>

lemonadern committed:
test(client): isolate invalid embedded sql fixture

lemonadern committed:
test(client): stop relying on transitive glob

lemonadern committed:
style: apply prettier formatting

tanzaku commented on client/src/test/helper.ts:
```diff
@@ -3,29 +3,45 @@ import * as path from "path";
 }
 
-async function sleep(ms: number) {
+export async function reopenDocument(
```
軽微な点ですが、現時点で未使用のようなので一旦削除して必要になったら作成するのはどうでしょうか

lemonadern replied:
ありがとうございます。いくつか残ってしまっていたものを整理しました。

https://github.com/future-architect/vscode-uroborosql-fmt/pull/99/commits/087d3ec861d4157fba060ecaa2e73d8883255675

lemonadern committed:
refactor(test): remove unused helpers

ppputtyo commented on client/src/command.ts:
```diff
@@ -222,3 +254,156 @@ export const buildImportSettingsFunction =
+    if (!selectionsMustNotOverlap(document, selectionRanges)) {
+      window.showErrorMessage(OVERLAPPING_SELECTIONS_MESSAGE);
+      return;
+    }
```
非常に軽微ですが、`!selectionsMustNotOverlap(document, selectionRanges)` で二重否定になっているので解消したいです。

`selectionsMustNotOverlap` の利用箇所はここの1箇所だけなので論理を反転するのはいかがでしょうか？

```diff
-const selectionsMustNotOverlap = (
+const hasOverlappingSelections = (
   document: TextDocument,
   selections: readonly Selection[],
 ): boolean => {
   const sortedRanges = selections
     .map(selectionToRange)
     .map((range) => ({
       start: document.offsetAt(range.start),
       end: document.offsetAt(range.end),
     }))
     .sort((left, right) => left.start - right.start);

   for (let i = 1; i < sortedRanges.length; i += 1) {
     if (sortedRanges[i].start < sortedRanges[i - 1].end) {
-      return false;
+      return true;
     }
   }

-  return true;
+  return false;
 };

-if (!selectionsMustNotOverlap(document, selectionRanges)) {
+if (hasOverlappingSelections(document, selectionRanges)) {
   window.showErrorMessage(OVERLAPPING_SELECTIONS_MESSAGE);
   return;
 }
```

lemonadern replied:
ありがとうございます。提案いただいた方針に変更しました。
https://github.com/future-architect/vscode-uroborosql-fmt/pull/99/commits/9c168d7033eded394dba03e6b3fe1471fbf1e73e

lemonadern committed:
Refine overlapping selection predicate naming

lemonadern committed:
test: add lint e2e coverage

lemonadern committed:
fix: exclude lint config path from formatter config sync

lemonadern committed:
docs: update README.md

lemonadern committed:
test: stabilize extension e2e runner

lemonadern committed:
test: add multi-root e2e suite

lemonadern committed:
refactor: tidy up tests (#116)

* test: revise vscode e2e suites

* test: use event-driven e2e waits

* refactor: simplify e2e test helpers

* test: slim down vscode e2e coverage

* refactor: narrow e2e helper exports

* refactor: simplify e2e test helpers and runners

* test: tidy up vscode e2e suites and fixtures

- remove unused lint fixtures and dead overrides/ignore config
- drop redundant env exports from e2e shell scripts
- add timeout context to helper waits and inline single-use stable helper
- share a Mocha runner and co-locate the multi-root index with its suite
- add a symmetric project-a multi-root test so its fixture/config is asserted
- stabilize the explicit-config lint test against transient default diagnostics

lemonadern commented:
e2e 基盤を先に main に追加してから移行をする
レビューコストの低減と、回帰検証の自動化のため

lemonadern commented:
#121 へ移行
